src/main/scala/com/fulcrumgenomics/fastq/DemuxFastqs.scala

```diff
@@ -0,0 +1,663 @@
+          case None => unreachable("sample barcode should have been set.")
+        }
+      }
+      // NB: we could check that no sample barcodes are reverse complements of each other.
+    }
+
+    // Determine the quality format
+    determineQualityFormat()
+
+    // Create the files for writing
+    val writers = createSamFileWriters(sampleSheet)
+
+    // Open the input FASTQs for reading
+    val nonEmptyFastqs = fastqs.filter(_.nonEmpty)
+    val reader = new ParallelFastqReader(nonEmptyFastqs.head, nonEmptyFastqs.tail:_*)
+
+    // Read in the quadruples of FASTQ records
+    val fastqConverter = new FastqConverter(
+      reader,
+      writers.map(_.getFileHeader),
+      sampleSheet,
+      umiTag,
+      multiUmiTags,
+      qualityFormat.getOrElse(unreachable("Quality format not set")),
+      minQ,
+      maxQ,
+      maxMismatches,
+      minMismatchDelta,
+      maxNoCalls,
+      matchReverseComplement,
+      readStructures:_*)
+
+    val progress = new com.fulcrumgenomics.util.ProgressLogger(logger)
+    fastqConverter.foreach { assignedRec =>
+      val writer = writers(assignedRec.sampleOrdinal-1)
+      assignedRec.records.foreach { rec => writer.addAlignment(rec) }
+      progress.record(assignedRec.records:_*)
+    }
+
+    // close them all
+    reader.close()
+    writers.foreach(_.close())
+
+    // write the metrics
+    Metric.write[SampleBarcodeMetric](metrics=fastqConverter.getBarcodeMetrics.toSeq, path=metrics)
+  }
+
+  private def determineQualityFormat(): Unit = {
+    val readers = fastq1.map { fastq => new FastqReader(fastq.toFile, allowAndIgnoreEmptyLines) }
+    val detector: QualityEncodingDetector = new QualityEncodingDetector
+    detector.add(QualityEncodingDetector.DEFAULT_MAX_RECORDS_TO_ITERATE, readers:_*)
+    readers.foreach(_.close())
+    val format = detector.generateBestGuess(QualityEncodingDetector.FileContext.FASTQ, qualityFormat.orNull)
+    if (detector.isDeterminationAmbiguous) {
+      logger.warning("Making ambiguous determination about fastq's quality encoding; more than one format possible based on observed qualities.")
+    }
+    logger.info(String.format("Auto-detected quality format as: %s.", format))
+    this.qualityFormat = Some(format)
+  }
+
+  /** Creates SAM file writers for each sample in the Sample Sheet, returning them in the same order as specified
+    * by each sample's ordinal.  An extra writer is appended for records that do not match a sample barcode. */
+  private def createSamFileWriters(sampleSheet: SampleSheet): List[SAMFileWriter] = {
+    val writers: List[SAMFileWriter] = sampleSheet.map { sample =>
+      val outputBam: Path       = DemuxFastqs.sampleOutputBam(this.output, sample)
+      val header: SAMFileHeader = createSamFileHeader(sample)
+      val writer: SAMFileWriter = new SAMFileWriterFactory().makeSAMOrBAMWriter(header, false, outputBam.toFile)
```

Comment:
Can drop the assignment and the next line, and just yield the value of `new ...`